Before you turn this homework in, make sure everything runs as expected. First, **restart the kernel** and then **run all cells** (in the menubar, select Kernel$\rightarrow$Restart \& Run all). If you get error messages, the grader will get the same errors and won't be able to grade your work. 

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE". You may delete the line(s) that say "**raise** NotImplementedError()" and replace them with your actual work. Write all your work in the cells provided in this notebook. **DO NOT** add or delete any cells. 

Please fill in your name in the cell below.

In [1]:
NAME = "Kevin Pham"

---

## Problem 1

The file "Temp.txt" which you can download from the homework tab on Canvas, contains the monthly high-temperatures (in degrees Celsius) at Heathrow airport for the years 1948 through 2016. Save the file on your computer (in a folder that you can find again). Run the code below, changing the path to the appropriate path on **your** computer. It's ok to use forward slashes in the path name, regardless of whether you're working on Windows or MacOS. 

In [2]:
# COMPLETE

# create empty list for data input
temperatures = []

from pathlib import Path

# set the appropriate path. Change "\" into "/" to correct pathing syntax
data_folder = Path('/Users/Kevin/Documents/SJSU/Spring 2023/Math 167PS/Homeworks')
# make sure to change this path to the actual folder where YOU have saved the data file. 

file_to_open = data_folder / "Temp.txt" # selection of file to open in the directed folder

file_to_open
# open the actual file
f = open(file_to_open)

# queue all data into the empty list declared earlier
for row in f:
    temperatures.append(float(row)) # queue all data in sequential order and all entries queued are float values



You have now created an object called temperatures. 

Answer the following questions by inserting your code into the code box (after the "# your code here" prompt. Comment out or delete the "raise NotImplementedError()") and answering the question in the form of a complete English sentence in the respective Markdown box. 

a) What kind of object type is "temperatures"?

In [3]:
# COMPLETE 
# YOUR CODE HERE
print(type(temperatures))
# raise NotImplementedError()

<class 'list'>


temperatures is a 'list' object

b) Describe what the "float()" and "append()" commands do, respectively, that have been applied above. Try to leave out the "float()" command. What happens then? 

float() - Changes the datatype of a number into float type numerical value

append() - Adds items to the end of a list

c) How many temperatures have been recorded in the file?

In [4]:
# COMPLETE
# YOUR CODE HERE
print(len(temperatures))
# raise NotImplementedError()

828


828 total entries

d) What are the highest and lowest temperatures recorded?

In [5]:
# COMPLETE
# YOUR CODE HERE
print(max(temperatures)) # finding max and print
print(min(temperatures)) # finding min and print
# raise NotImplementedError()

28.2
0.8


max: 28.2
min: 0.8

e) The first observation in the `temperatures` object is from a January 1948. Using only the commands you learned from chapter 5 (that is, using base python and no additional python libraries), find the the mean January high temperature at London Heathrow between the years 1948 and 2016. Have your code round your answer to two decimal digits accuracy and return the answer as a float.  

In [6]:
# COMPLETE
# YOUR CODE HERE
jan = temperatures[0:828:12] # list of just january temperatures
avg = sum(jan)/len(jan) # take the average of the january temperatures
print(round(avg,2)) # list of just jan. between 1948 - 2016 avg rounded to 2 decimal points
# raise NotImplementedError()

7.56


f) Find the median monthly high temperature (using only the commands you learned from chapter 5 and base python, not using additional modules or libraries). Use Python to round your result to two after-decimal digits. State the result and explain briefly how you find the median in the markdown box below. 

In [7]:
# COMPLETE
# YOUR CODE HERE
tempSort = sorted(temperatures) # sort them from descending to ascending

# find median position
midIndex = (len(temperatures)-1)//2
# check if it's even or odd
if(midIndex % 2):
    print(tempSort[midIndex]) # odd # of elements
else:
    print((tempSort[midIndex] + tempSort[midIndex+1])/2) # even # of elements
# raise NotImplementedError()

14.7


Median: 14.7

## Problem 2 

Much of data science these days is based on text data. The text to be analyzed may be literature or postings on social media. The tasks to be performed could be sentiment analysis (trying to guess whether the writing is "pro" or "con" on a certain issue) to automatically detecting postings by Malicious bots. However, before you can perform any kind of machine learning procedure on a text, you have to first learn how to "clean" the text. That usually means removing all non-text elements and sometimes also removing filler words such as "the", "it", "and" etc. After that, what is usually analyzed is the frequency with which certain words appear and most often actually not the *order* in which they appear. 

The text below are the first three paragraphs of Franz Kafka's book "Metamorphosis". Our goal is to clean this text passage. 

In [8]:
text = """One morning, when Gregor Samsa woke from troubled dreams, he found
himself transformed in his bed into a horrible vermin.  He lay on
his armour-like back, and if he lifted his head a little he could
see his brown belly, slightly domed and divided by arches into stiff
sections.  The bedding was hardly able to cover it and seemed ready
to slide off any moment.  His many legs, pitifully thin compared
with the size of the rest of him, waved about helplessly as he
looked.

"What's happened to me?" he thought.  It wasn't a dream.  His room,
a proper human room although a little too small, lay peacefully
between its four familiar walls.  A collection of textile samples
lay spread out on the table - Samsa was a travelling salesman - and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice, gilded frame.  It showed
a lady fitted out with a fur hat and fur boa who sat upright,
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer.

Gregor then turned to look out the window at the dull weather.
Drops of rain could be heard hitting the pane, which made him feel
quite sad.  "How about if I sleep a little bit longer and forget all
this nonsense", he thought, but that was something he was unable to
do because he was used to sleeping on his right, and in his present
state couldn't get into that position.  However hard he threw
himself onto his right, he always rolled back to where he was.  He
must have tried it a hundred times, shut his eyes so that he
wouldn't have to look at the floundering legs, and only stopped when
he began to feel a mild, dull pain there that he had never felt
before."""

a) Look at the text. What non-alphabetic character elements do you see in the text?

non-alphanumerics: " ' . , - ?

b) Using only the commands covered in class so far (in base python without use of additional libraries), write code to remove all non-alphabetic characters (such as punctuation or hyphens).

In [9]:
# COMPLETE
# YOUR CODE HERE

# remove "
text = text.replace('"',"")

# remove '
text = text.replace("'","")

# remove - 
text = text.replace("-","")

# remove ?
text = text.replace("?","")

# remove .
text = text.replace(".","")

# remove ,
text = text.replace(",","")

print(text) # print out without any non-alphanumerics
# raise NotImplementedError()

One morning when Gregor Samsa woke from troubled dreams he found
himself transformed in his bed into a horrible vermin  He lay on
his armourlike back and if he lifted his head a little he could
see his brown belly slightly domed and divided by arches into stiff
sections  The bedding was hardly able to cover it and seemed ready
to slide off any moment  His many legs pitifully thin compared
with the size of the rest of him waved about helplessly as he
looked

Whats happened to me he thought  It wasnt a dream  His room
a proper human room although a little too small lay peacefully
between its four familiar walls  A collection of textile samples
lay spread out on the table  Samsa was a travelling salesman  and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice gilded frame  It showed
a lady fitted out with a fur hat and fur boa who sat upright
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer

Gregor 

c) Write code to make everything lower case and convert the text into a list of single words. 

In [10]:
# COMPLETE
# YOUR CODE HERE
textList = text.lower().split()
print(textList) # ordering methods matter. Make all the words into lowercase first then convert to a list
# raise NotImplementedError()

['one', 'morning', 'when', 'gregor', 'samsa', 'woke', 'from', 'troubled', 'dreams', 'he', 'found', 'himself', 'transformed', 'in', 'his', 'bed', 'into', 'a', 'horrible', 'vermin', 'he', 'lay', 'on', 'his', 'armourlike', 'back', 'and', 'if', 'he', 'lifted', 'his', 'head', 'a', 'little', 'he', 'could', 'see', 'his', 'brown', 'belly', 'slightly', 'domed', 'and', 'divided', 'by', 'arches', 'into', 'stiff', 'sections', 'the', 'bedding', 'was', 'hardly', 'able', 'to', 'cover', 'it', 'and', 'seemed', 'ready', 'to', 'slide', 'off', 'any', 'moment', 'his', 'many', 'legs', 'pitifully', 'thin', 'compared', 'with', 'the', 'size', 'of', 'the', 'rest', 'of', 'him', 'waved', 'about', 'helplessly', 'as', 'he', 'looked', 'whats', 'happened', 'to', 'me', 'he', 'thought', 'it', 'wasnt', 'a', 'dream', 'his', 'room', 'a', 'proper', 'human', 'room', 'although', 'a', 'little', 'too', 'small', 'lay', 'peacefully', 'between', 'its', 'four', 'familiar', 'walls', 'a', 'collection', 'of', 'textile', 'samples', 'l

d) How many total words are in the list (if armour-like is considered to be two words)? Count, how often the word "he" appears in the three paragraphs. Don't count if "he" is part of another word, e.g., "tHEn". 

In [11]:
# COMPLETE
# YOUR CODE HERE

# total words in the list
print(len(textList))

print(textList.count("he")) # find all instances of "he"

# raise NotImplementedError()

313
17


313 total words in the list
17 instances of "he"

e) How many distinct (different) words appear in the first three paragragraphs?

In [12]:
# COMPLETE
# YOUR CODE HERE
unique = set(textList) # find only the unique words in the paragraph from the list
print(len(unique))
# raise NotImplementedError()

194


194 unique words in the string